server/src/main/java/org/opensearch/search/DefaultSearchContext.java

```diff
@@ -1056,40 +1058,54 @@ public BucketCollectorProcessor bucketCollectorProcessor() {
+     * @return the resolved concurrent segment search mode: "none", "auto", or "all"
      */
     private String evaluateConcurrentSearchMode(Executor concurrentSearchExecutor) {
-        // Do not use concurrent segment search for system indices or throttled requests. See:
-        // https://github.com/opensearch-project/OpenSearch/issues/12951
-        if (indexShard.isSystem() || indexShard.indexSettings().isSearchThrottled()) {
+        // Skip concurrent search for system indices, throttled requests, or if dependencies are missing
+        if (indexShard.isSystem()
+            || indexShard.indexSettings().isSearchThrottled()
+            || clusterService == null
+            || concurrentSearchExecutor == null) {
             return CONCURRENT_SEGMENT_SEARCH_MODE_NONE;
         }
-        if ((clusterService != null) && concurrentSearchExecutor != null) {
-            String concurrentSearchMode = indexService.getIndexSettings()
-                .getSettings()
-                .get(
-                    IndexSettings.INDEX_CONCURRENT_SEGMENT_SEARCH_MODE.getKey(),
-                    clusterService.getClusterSettings().getOrNull(CLUSTER_CONCURRENT_SEGMENT_SEARCH_MODE)
-                );
-            if (concurrentSearchMode != null) {
-                return concurrentSearchMode;
-            }
 
-            // mode setting not set, fallback to concurrent_segment_search.enabled setting
-            return indexService.getIndexSettings()
-                .getSettings()
-                .getAsBoolean(
-                    IndexSettings.INDEX_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(),
-                    clusterService.getClusterSettings().get(CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING)
-                ) ? CONCURRENT_SEGMENT_SEARCH_MODE_ALL : CONCURRENT_SEGMENT_SEARCH_MODE_NONE;
+        Settings indexSettings = indexService.getIndexSettings().getSettings();
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+
+        // Evaluate legacy boolean setting (deprecated):
+        // - true → "all"
+        // - false → "none"
+        // If defined, this overrides the newer string-based setting.
+        final Boolean concurrentSearchLegacySetting = indexSettings.getAsBoolean(
+            IndexSettings.INDEX_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(),
+            clusterSettings.getOrNull(CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING)
+        );
+
+        if (concurrentSearchLegacySetting != null) {
+            return concurrentSearchLegacySetting ? CONCURRENT_SEGMENT_SEARCH_MODE_ALL : CONCURRENT_SEGMENT_SEARCH_MODE_NONE;
         }
-        return CONCURRENT_SEGMENT_SEARCH_MODE_NONE;
+
+        // Fallback to string-based setting: "none", "auto", or "all"
+        return indexSettings.get(
+            IndexSettings.INDEX_CONCURRENT_SEGMENT_SEARCH_MODE.getKey(),
+            clusterSettings.get(CLUSTER_CONCURRENT_SEGMENT_SEARCH_MODE)
+        );
```

Comment:
There is one other case which we will need to handle. When both new and old setting is explicitly set, then we need to give precedence to new setting. If both are not set then use the default. If either of them is explicitly set then use the one which is set.